src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -28,34 +29,50 @@ constexpr uint64_t kCodeScratchPadSize = 1024;
 constexpr const char* kLibdlSoname = "libdl.so.2";
-constexpr const char* kDlopenInLibdl = "dlopen";
-constexpr const char* kDlopenInLibc = "__libc_dlopen_mode";
-constexpr const char* kDlsymInLibdl = "dlsym";
-constexpr const char* kDlsymInLibc = "__libc_dlsym";
-constexpr const char* kDlcloseInLibdl = "dlclose";
-constexpr const char* kDlcloseInLibc = "__libc_dlclose";
+
+// Represents a symbol (function) in a module.
+// The member variables are string views as these are meant to be kept as constexpr constants.
+struct FunctionLocatorView {
+  std::string_view module_name;
+  std::string_view function_name;
+};
+
+constexpr FunctionLocatorView kDlopenInLibdl{kLibdlSoname, "dlopen"};
+constexpr FunctionLocatorView kDlopenInLibc{kLibcSoname, "dlopen"};
+constexpr FunctionLocatorView kDlopenFallbackInLibc{kLibcSoname, "__libc_dlopen_mode"};
+
+constexpr FunctionLocatorView kDlsymInLibdl{kLibdlSoname, "dlsym"};
+constexpr FunctionLocatorView kDlsymInLibc{kLibcSoname, "dlsym"};
+constexpr FunctionLocatorView kDlsymFallbackInLibc{kLibcSoname, "__libc_dlsym"};
+
+constexpr FunctionLocatorView kDlcloseInLibdl{kLibdlSoname, "dlclose"};
+constexpr FunctionLocatorView kDlcloseInLibc{kLibcSoname, "dlclose"};
+constexpr FunctionLocatorView kDlcloseFallbackInLibc{kLibcSoname, "__libc_dlclose"};
 
 // Returns the absolute virtual address of a function in a module of a process as
-// FindFunctionAddress does but accepts a fallback symbol if the primary one cannot be resolved.
-ErrorMessageOr<uint64_t> FindFunctionAddressWithFallback(pid_t pid, std::string_view module,
-                                                         std::string_view function,
-                                                         std::string_view fallback_module,
-                                                         std::string_view fallback_function) {
-  ErrorMessageOr<uint64_t> primary_address_or_error = FindFunctionAddress(pid, module, function);
-  if (primary_address_or_error.has_value()) {
-    return primary_address_or_error.value();
-  }
-  ErrorMessageOr<uint64_t> fallback_address_or_error =
-      FindFunctionAddress(pid, fallback_module, fallback_function);
-  if (fallback_address_or_error.has_value()) {
-    return fallback_address_or_error.value();
+// FindFunctionAddress does but accepts a list of module- and function names and returns
+// the address of the first found function.
+ErrorMessageOr<uint64_t> FindFunctionAddressWithFallback(
+    pid_t pid, absl::Span<const FunctionLocatorView> function_locators) {
+  std::string error_message;
+
+  for (const auto& function_locator : function_locators) {
+    ErrorMessageOr<uint64_t> address_or_error =
+        FindFunctionAddress(pid, function_locator.module_name, function_locator.function_name);
+    if (address_or_error.has_value()) {
+      return address_or_error.value();
+    }
+
+    if (!error_message.empty()) {
+      error_message.append(" - Continuing to look for the next best option...\n");
```

Comment:
If I understand correctly this error message could go all the way to the client. This string seems more fitting for something that you would read live in a terminal rather than a finished error message.

In case of multiple errors, could we have something like the following?
```
Failed to load symbol "%s" from module "%s" with error: "%s"
Also failed to load fallback symbol "%s" from module "%s" with error: "%s"
```

Reply:
Yes. Works for me as well!

Reply:
Done. Thx.